File: packages/dart/lib/src/workspace/progress.dart
import '../task/model.dart';
import '../workflow/model.dart';
import 'model.dart';

/// 工作区聚合 / 流水判定：走过哪几步、下一步是哪。
///
/// 「走过哪几步」从任务流水读出：带后缀的流水（`·审` / `·判`）给这一步**投票**，
/// 不带后缀的（重走一遍）**把结论从头算**。工作流从工作区里按名字取。
/// 出处：`docs/specification/process/task.md`·读法。
extension WorkspaceProgress on Workspace {
  /// 这件任务走过哪几步（按流水里的先后）。
  List<String> doneSteps(Task task) {
    final flow = workflow(task.workflowName);
    if (flow == null) return const [];
    return _doneSteps(task, flow);
  }

  /// 这件任务下一个没走到的步骤（按定义里的顺序）；都走过或没有定义给 `null`。
  String? nextStep(Task task) {
    final flow = workflow(task.workflowName);
    if (flow == null) return null;
    return _nextStep(task, flow);
  }

  /// 状态行：下一步是谁，或者都走过了。
  String stateLine(Task task) {
    final flow = workflow(task.workflowName);
    if (flow == null) return '工作区里没有这条工作流：${task.workflowName}';
    return _stateLine(task, flow);
  }
}

/// 走过哪几步：工作流上没有的步骤名不算数。
List<String> _doneSteps(Task task, Workflow workflow) {
  final names = workflow.stepNames;
  final verdict = <String, bool>{};
  final order = <String>[];
  for (final event in task.journal) {
    final at = event.step.indexOf('·');
    final step = at < 0 ? event.step : event.step.substring(0, at);
    if (!names.contains(step)) continue;
    if (verdict.containsKey(step)) {
      verdict[step] = at >= 0 ? (verdict[step]! && event.ok) : event.ok;
    } else {
      verdict[step] = event.ok;
      order.add(step);
    }
  }
  return order.where((step) => verdict[step] == true).toList();
}

/// 第一个没走到的步骤（按工作流里的顺序）。
String? _nextStep(Task task, Workflow workflow) {
  final finished = _doneSteps(task, workflow);
  for (final name in workflow.stepNames) {
    if (!finished.contains(name)) return name;
  }
  return null;
}

/// 状态行：下一步是谁，或者都走过了。
String _stateLine(Task task, Workflow workflow) {
  final names = workflow.stepNames;
  if (names.isEmpty) {
    return '这条工作流没有步骤：${workflow.name}';
  }
  final next = _nextStep(task, workflow);
  return next == null ? '${names.length} 个步骤都走过了' : '下一步：$next';
}

File: packages/rust/src/workspace/progress.rs
//! 工作区聚合 / 流水判定：走过哪几步、下一步是哪。
//!
//! 「走过哪几步」从任务流水读出：带后缀的流水（`·审` / `·判`）给这一步**投票**，
//! 不带后缀的（重走一遍）**把结论从头算**。工作流从工作区里按名字取。
//! 出处：`docs/specification/process/task.md`·读法。

use super::model::Workspace;
use crate::task::Task;
use crate::workflow::Workflow;

impl Workspace {
    /// 这件任务走过哪几步（按流水里的先后）。
    pub fn done_steps(&self, task: &Task) -> Vec<String> {
        match self.workflow(&task.workflow_name) {
            Some(workflow) => done_steps(task, workflow),
            None => Vec::new(),
        }
    }

    /// 这件任务下一个没走到的步骤（按定义里的顺序）；都走过或没有定义给 `None`。
    pub fn next_step(&self, task: &Task) -> Option<String> {
        let workflow = self.workflow(&task.workflow_name)?;
        next_step(task, workflow)
    }

    /// 状态行：下一步是谁，或者都走过了。
    pub fn state_line(&self, task: &Task) -> String {
        let Some(workflow) = self.workflow(&task.workflow_name) else {
            return format!("工作区里没有这条工作流：{}", task.workflow_name);
        };
        state_line(task, workflow)
    }
}

/// 走过哪几步：工作流上没有的步骤名不算数。
fn done_steps(task: &Task, workflow: &Workflow) -> Vec<String> {
    let names = workflow.step_names();
    let mut verdict: Vec<(String, bool)> = Vec::new();
    for event in &task.journal {
        let (step, extra) = match event.step.split_once('·') {
            Some((step, rest)) => (step.to_string(), Some(rest.to_string())),
            None => (event.step.clone(), None),
        };
        if !names.contains(&step) {
            continue;
        }
        match verdict.iter_mut().find(|(name, _)| *name == step) {
            Some((_, last)) => {
                if extra.is_some() {
                    *last = *last && event.ok;
                } else {
                    *last = event.ok;
                }
            }
            None => verdict.push((step, event.ok)),
        }
    }
    verdict
        .into_iter()
        .filter(|(_, ok)| *ok)
        .map(|(name, _)| name)
        .collect()
}

/// 第一个没走到的步骤（按工作流里的顺序）。
fn next_step(task: &Task, workflow: &Workflow) -> Option<String> {
    let finished = done_steps(task, workflow);
    workflow
        .step_names()
        .into_iter()
        .find(|name| !finished.contains(name))
}

/// 状态行：下一步是谁，或者都走过了。
fn state_line(task: &Task, workflow: &Workflow) -> String {
    let names = workflow.step_names();
    if names.is_empty() {
        return format!("这条工作流没有步骤：{}", workflow.name);
    }
    match next_step(task, workflow) {
        Some(step) => format!("下一步：{step}"),
        None => format!("{} 个步骤都走过了", names.len()),
    }
}
